From d8ddd2c51f7e9934e685df8ebc703223ee794cf7 Mon Sep 17 00:00:00 2001
From: Chen Wandun <chenwandun@huawei.com>
Date: Sat, 12 Nov 2022 15:10:28 +0800
Subject: [PATCH 0030/1398] [Huawei] mm: add config isolation for
 per-memcg-reclaim

Offering: HULK
hulk inclusion
category: feature
bugzilla: 187966

-------------------------------

Add CONFIG_MEMCG_RECLAIM to separate feature of per-memcg-reclaim
from baseline.

Signed-off-by: Chen Wandun <chenwandun@huawei.com>
---
 arch/arm64/configs/openeuler_defconfig |  1 +
 arch/x86/configs/openeuler_defconfig   |  1 +
 mm/Kconfig                             | 10 ++++++++++
 mm/memcontrol.c                        |  4 ++++
 4 files changed, 16 insertions(+)

diff --git a/arch/arm64/configs/openeuler_defconfig b/arch/arm64/configs/openeuler_defconfig
index f7a86bcd9b5c..2cd41b103121 100644
--- a/arch/arm64/configs/openeuler_defconfig
+++ b/arch/arm64/configs/openeuler_defconfig
@@ -1077,6 +1077,7 @@ CONFIG_CLEANCACHE=y
 CONFIG_FRONTSWAP=y
 CONFIG_SHRINK_PAGECACHE=y
 CONFIG_MEMCG_QOS=y
+CONFIG_MEMCG_RECLAIM=y
 CONFIG_ETMEM_SCAN=m
 CONFIG_ETMEM_SWAP=m
 CONFIG_ETMEM=y
diff --git a/arch/x86/configs/openeuler_defconfig b/arch/x86/configs/openeuler_defconfig
index db64ad678f77..812b061ff1da 100644
--- a/arch/x86/configs/openeuler_defconfig
+++ b/arch/x86/configs/openeuler_defconfig
@@ -1024,6 +1024,7 @@ CONFIG_CLEANCACHE=y
 CONFIG_FRONTSWAP=y
 CONFIG_SHRINK_PAGECACHE=y
 CONFIG_MEMCG_QOS=y
+CONFIG_MEMCG_RECLAIM=y
 CONFIG_ETMEM_SCAN=m
 CONFIG_ETMEM_SWAP=m
 CONFIG_ETMEM=y
diff --git a/mm/Kconfig b/mm/Kconfig
index 1ff990c1b894..eef871ccf465 100644
--- a/mm/Kconfig
+++ b/mm/Kconfig
@@ -519,6 +519,16 @@ config MEMCG_QOS
 
 	  If unsure, say "n".
 
+config MEMCG_RECLAIM
+	bool "Enable Per Memory Cgroup Reclaim in Cgroup V1"
+	depends on MEMCG
+	default n
+	help
+	  MEMCG_RECLAIM add memory.reclaim file in each memcg, memory reclaim
+	  would be triggered when writing to this file.
+
+	  If unsure, say "n".
+
 config ETMEM_SCAN
 	tristate "module: etmem page scan for etmem support"
 	depends on ETMEM
diff --git a/mm/memcontrol.c b/mm/memcontrol.c
index df62715b521b..a38b1196331d 100644
--- a/mm/memcontrol.c
+++ b/mm/memcontrol.c
@@ -5459,10 +5459,14 @@ static struct cftype mem_cgroup_legacy_files[] = {
 		.file_offset = offsetof(struct mem_cgroup, events_local_file),
 		.seq_show = memcg_events_local_show,
 	},
+
+#ifdef CONFIG_MEMCG_RECLAIM
 	{
 		.name = "reclaim",
 		.write = memory_reclaim,
 	},
+#endif
+
 #ifdef CONFIG_CGROUP_V1_WRITEBACK
 	{
 		.name = "wb_blkio_ino",
-- 
2.17.1

